sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -0,0 +1,216 @@
+    val maxCmp = ordering.compare(value, maxInToType)
+
+    if (maxCmp > 0) {
+      exp match {
+        case EqualTo(_, _) | GreaterThan(_, _) | GreaterThanOrEqual(_, _) =>
+          fromExp.falseIfNotNull
+          // falseIfNotNull(fromExp)
+        case LessThan(_, _) | LessThanOrEqual(_, _) =>
+          fromExp.trueIfNotNull
+        case EqualNullSafe(_, _) =>
+          // make sure the expression is evaluated if it is non-deterministic
+          if (exp.deterministic) FalseLiteral else exp
+        case _ => exp // impossible but safe guard, same below
+      }
+    } else if (maxCmp == 0) {
+      exp match {
+        case GreaterThan(_, _) =>
+          fromExp.falseIfNotNull
+        case LessThanOrEqual(_, _) =>
+          fromExp.trueIfNotNull
+        case LessThan(_, _) =>
+          Not(EqualTo(fromExp, Literal(max, fromType)))
+        case GreaterThanOrEqual(_, _) | EqualTo(_, _) | EqualNullSafe(_, _) =>
+          EqualTo(fromExp, Literal(max, fromType))
+        case _ => exp
+      }
+    } else if (minCmp < 0) {
+      exp match {
+        case GreaterThan(_, _) | GreaterThanOrEqual(_, _) =>
+          fromExp.trueIfNotNull
+        case LessThan(_, _) | LessThanOrEqual(_, _) | EqualTo(_, _) =>
+          fromExp.falseIfNotNull
+        case EqualNullSafe(_, _) =>
+          // make sure the expression is evaluated if it is non-deterministic
+          if (exp.deterministic) FalseLiteral else exp
+        case _ => exp
+      }
+    } else if (minCmp == 0) {
+      exp match {
+        case LessThan(_, _) =>
+          fromExp.falseIfNotNull
+        case GreaterThanOrEqual(_, _) =>
+          fromExp.trueIfNotNull
+        case GreaterThan(_, _) =>
+          Not(EqualTo(fromExp, Literal(min, fromType)))
+        case LessThanOrEqual(_, _) | EqualTo(_, _) | EqualNullSafe(_, _) =>
+          EqualTo(fromExp, Literal(min, fromType))
+        case _ => exp
+      }
+    } else {
+      // This means `value` is within range `(min, max)`. Optimize this by moving the cast to the
+      // literal side.
+      val lit = Cast(Literal(value), fromType)
+      exp match {
+        case GreaterThan(_, _) => GreaterThan(fromExp, lit)
+        case GreaterThanOrEqual(_, _) => GreaterThanOrEqual(fromExp, lit)
+        case EqualTo(_, _) => EqualTo(fromExp, lit)
+        case EqualNullSafe(_, _) => EqualNullSafe(fromExp, lit)
+        case LessThan(_, _) => LessThan(fromExp, lit)
+        case LessThanOrEqual(_, _) => LessThanOrEqual(fromExp, lit)
+        case _ => exp
+      }
+    }
+  }
+
+  /**
+   * Check if the input `fromExp` can be safely cast to `toType` without any loss of precision,
+   * i.e., the conversion is injective. Note this only handles the case when both sides are of
+   * integral type.
+   */
+  private def canImplicitlyCast(fromExp: Expression, toType: DataType): Boolean = {
+    fromExp.dataType.isInstanceOf[IntegralType] && toType.isInstanceOf[IntegralType] &&
+    Cast.canUpCast(fromExp.dataType, toType)
```

Comment:
Is this check still required? We check it in the pattern matching for `toType.isInstanceOf[IntegralType]`. For Cast.canUpCast, seems this will always be true since it's already in the cast expression constructed by Spark, right?

Reply:
No, it is not required but just to make the method self-complete. 

> For Cast.canUpCast, seems this will always be true since it's already in the cast expression constructed by Spark

Hmm I'm not sure. Users can say `cast(<some_long_column> as int) > 42` and we'd want to skip this case, I think?